File: 01311. Get Watched Videos by Your Friends/solution.cs
/**
 * Problem: 1311. Get Watched Videos by Your Friends
 *
 * Difficulty: Medium
 *
 * Language: C#
 *
 * Performance: Runtime - 39 ms (Beats 100%)
 */

public class Solution
{
  public IList<string> WatchedVideosByFriends(
    IList<IList<string>> watchedVideos,
    int[][] friends,
    int id,
    int level
  )
  {
    // Initialize queue for BFS traversal
    Queue<int> queue = new Queue<int>();

    // Add starting person to queue
    queue.Enqueue(id);

    // Track visited people to avoid cycles
    bool[] visited = new bool[friends.Length];

    // Mark starting person as visited
    visited[id] = true;

    // Perform BFS to reach target level
    for (int currentLevel = 0; currentLevel < level; currentLevel++)
    {
      // Number of people at current level to process
      int levelSize = queue.Count;

      // Process all people at current level
      for (int i = 0; i < levelSize; i++)
      {
        int person = queue.Dequeue();

        // Add all unvisited friends to next level
        foreach (int friend in friends[person])
        {
          if (!visited[friend])
          {
            queue.Enqueue(friend);
            visited[friend] = true;
          }
        }
      }
    }

    // Count frequency of videos from friends at target level
    Dictionary<string, int> frequency = new Dictionary<string, int>();

    // Process each friend in the queue (all at target level)
    foreach (int person in queue)
    {
      foreach (var video in watchedVideos[person])
      {
        frequency[video] = frequency.GetValueOrDefault(video) + 1;
      }
    }

    // Get all video keys
    string[] result = frequency.Keys.ToArray();

    // Sort by frequency (ascending), then alphabetically
    Array.Sort(
      result,
      (x, y) =>
      {
        // Compare by frequency first
        int compare = frequency[x] - frequency[y];

        // If frequencies differ, return frequency comparison
        if (compare != 0)
          return compare;

        // If frequencies are equal, compare alphabetically
        return string.Compare(x, y, StringComparison.Ordinal);
      }
    );

    // Return sorted result
    return result;
  }
}

File: 01311. Get Watched Videos by Your Friends/solution.go
/**
 * Problem: 1311. Get Watched Videos by Your Friends
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func watchedVideosByFriends(watchedVideos [][]string, friends [][]int, id int, level int) []string {
	// Number of people in the network
	n := len(friends)

	// Track visited people to avoid cycles
	visited := make([]bool, n)

	// Initialize queue with starting person
	queue := []int{id}

	// Mark starting person as visited
	visited[id] = true

	// Perform BFS to reach target level
	for currentLevel := 0; currentLevel < level; currentLevel++ {
		// Number of people at current level
		levelSize := len(queue)

		// Process all people at current level
		for i := 0; i < levelSize; i++ {
			// Get current person from front of queue
			current := queue[0]
			queue = queue[1:]

			// Add all unvisited friends to next level
			for _, friend := range friends[current] {
				if !visited[friend] {
					visited[friend] = true
					queue = append(queue, friend)
				}
			}
		}
	}

	// Count frequency of videos from friends at target level
	frequency := make(map[string]int)

	// Process each friend remaining in queue (all at target level)
	for _, friend := range queue {
		for _, video := range watchedVideos[friend] {
			frequency[video]++
		}
	}

	// Create slice to store video-frequency pairs
	videos := make([]VideoFreq, 0, len(frequency))

	// Populate videos slice from frequency map
	for name, count := range frequency {
		videos = append(videos, VideoFreq{name, count})
	}

	// Sort by frequency (ascending), then alphabetically
	sort.Slice(videos, func(i, j int) bool {
		// If frequencies are equal, sort alphabetically
		if videos[i].Count == videos[j].Count {
			return videos[i].Name < videos[j].Name
		}

		// Otherwise sort by frequency
		return videos[i].Count < videos[j].Count
	})

	// Create result slice for video names
	result := make([]string, len(videos))

	// Extract only video names from sorted videos
	for i, video := range videos {
		result[i] = video.Name
	}

	// Return sorted video names
	return result
}

// Helper struct to store video name and frequency
type VideoFreq struct {
	Name  string
	Count int
}
